portalnet/src/overlay_service.rs

```diff
@@ -758,58 +758,146 @@ where
+                        tokio::spawn(async move {
+                            Self::process_received_content(
+                                kbuckets,
+                                command_tx,
+                                validator,
+                                store,
+                                (content.clone(), true),
+                                content_key,
+                                callback,
+                                query_info.trace,
+                                nodes_to_poke,
+                            );
+                        });
+                    }
+                    FindContentQueryResult::Utp {
+                        connection_id,
+                        peer,
+                        nodes_to_poke,
+                    } => {
+                        let metrics = self.metrics.clone();
+                        let utp = self.utp_socket.clone();
+                        let id = u16::from_be(connection_id);
+                        let source = match self.find_enr(&peer) {
+                            Some(enr) => enr,
+                            _ => {
+                                warn!("Received uTP payload from unknown peer");
+                                return;
+                            }
+                        };
+                        let cid = utp_rs::cid::ConnectionId {
+                            recv: id,
+                            send: id.wrapping_add(1),
+                            peer: crate::discovery::UtpEnr(source),
+                        };
+                        let validator = self.validator.clone();
+                        let store = self.store.clone();
+                        let kbuckets = self.kbuckets.clone();
+                        let command_tx = self.command_tx.clone();
+                        tokio::spawn(async move {
+                            metrics.report_utp_active_inc(UtpDirectionLabel::Inbound);
+                            let mut stream =
+                                match utp.connect_with_cid(cid.clone(), UTP_CONN_CFG).await {
+                                    Ok(stream) => stream,
+                                    Err(err) => {
+                                        metrics.report_utp_outcome(
+                                            UtpDirectionLabel::Inbound,
+                                            UtpOutcomeLabel::FailedConnection,
+                                        );
+                                        warn!(
+                                            %err,
+                                            cid.send,
+                                            cid.recv,
+                                            peer = ?cid.peer.client(),
+                                            "Unable to establish uTP conn based on Accept",
+                                        );
+                                        return;
```

Comment:
Just thinking this through: If someone offers content and then the connection step fails, we just stop trying to collect the content, right? It will time out eventually, I guess. I suppose we could at least "time out" right away so the caller can re-run the call if they want.

In an ideal world, we would probably continue asking other peers for the same content. 👍🏻 to put something like this in an issue, for follow-up work. Plus maybe a comment here and the other early returns, noting the current behavior.